pkg/juicefs/juicefs.go

```diff
@@ -216,6 +216,10 @@ func (fs *jfs) CreateVol(ctx context.Context, volumeID, subPath string) (string,
 }
 
 func (fs *jfs) BindTarget(ctx context.Context, bindSource, target string) error {
+	lock := config.GetBindLock(target)
+	lock.Lock()
+	defer lock.Unlock()
+
```

Comment:
this lock might not be necessary, only `NodePublishVolume` will call it, usually not concurrent calls to the same volume

Reply:
consider locking the volume during the [NodePublishVolume](https://github.com/juicedata/juicefs-csi-driver/blob/6ee30b401af8e8f9782e9556f4585c738b5e8a75/pkg/driver/node.go#L101) process.

like this

https://github.com/juicedata/juicefs-csi-driver/blob/6ee30b401af8e8f9782e9556f4585c738b5e8a75/pkg/driver/controller.go#L147-L150

Reply:
when NodePublishVolume call timeout, kubelet will retry call NodePublishVolume  (we have one pod use multi pvcs by deferent juicefs filesystems, it can trigger createOrAddRef lock hanging)  like this

![image](https://github.com/juicedata/juicefs-csi-driver/assets/6017354/725e2051-c21d-475d-9875-14d24b1c2f16)


I0522 01:40:12.170382       1 node.go:99] NodePublishVolume: creating dir /var/lib/kubelet/pods/0412266a-2a91-424b-a548-f40fedab48b6/volumes/kubernetes.io~csi/gitn-ci-cache-gitlab/mount
.......
I0522 01:46:15.370284       1 node.go:99] NodePublishVolume: creating dir /var/lib/kubelet/pods/0412266a-2a91-424b-a548-f40fedab48b6/volumes/kubernetes.io~csi/gitn-ci-cache-gitlab/mount
I0522 01:48:19.456360       1 node.go:99] NodePublishVolume: creating dir /var/lib/kubelet/pods/0412266a-2a91-424b-a548-f40fedab48b6/volumes/kubernetes.io~csi/gitn-ci-cache-gitlab/mount

**hanging**


I0522 02:28:32.757529       1 node.go:155] NodePublishVolume: mounted gitn-ci-cache-gitlab at /var/lib/kubelet/pods/0412266a-2a91-424b-a548-f40fedab48b6/volumes/kubernetes.io~csi/gitn-ci-cache-gitlab/mount with options []
.....
I0522 02:28:37.242308       1 node.go:155] NodePublishVolume: mounted gitn-ci-cache-gitlab at /var/lib/kubelet/pods/0412266a-2a91-424b-a548-f40fedab48b6/volumes/kubernetes.io~csi/gitn-ci-cache-gitlab/mount with options []
I0522 02:28:38.443522       1 node.go:155] NodePublishVolume: mounted gitn-ci-cache-gitlab at /var/lib/kubelet/pods/0412266a-2a91-424b-a548-f40fedab48b6/volumes/kubernetes.io~csi/gitn-ci-cache-gitlab/mount with options []

Reply:
This doesn't like concurrent calls, is multiple calls. this operation is currently idempotent

btw, the `NodePublishVolume` It may need to be locked, but not here, refer to the above reply

Reply:
NodePublishVolume 只有当超时时上层重试调用才会出现并发问题，（如果问题1的锁阻塞解决之后这里出现的概率会更低）， NodePublishVolume 有两处会产生并发问题，第一初Jmount的createOrAddRef已经加了pod锁解决了。第二处就是BindTarget 可能出现并行调用。 TryAcquire 是非阻塞锁，这里BindTarget需要阻塞等待。如果觉得这里不需要加锁的话也可以去掉，我这里的修改比较简单。第一个createOrAddRef lock锁阻塞的问题是需要修复的

Reply:
我们的生产环境因为第一个lock阻塞问题，导致大量的NodePublishVolume重试，然后我们因为用的是旧版本csi，导致重复挂载把mount info 用尽了，这里新版本基本上已经解决了重复挂载的问题了
https://github.com/juicedata/juicefs-csi-driver/issues/404